ndb/tests/unit/test_model.py

```diff
@@ -1704,9 +1705,69 @@ def test__from_base_type(self):
 
 class TestJsonProperty:
     @staticmethod
-    def test_constructor():
-        with pytest.raises(NotImplementedError):
-            model.JsonProperty()
+    def test_constructor_defaults():
+        prop = model.JsonProperty()
+        # Check that none of the constructor defaults were used.
+        assert prop.__dict__ == {}
+
+    @staticmethod
+    def test_constructor_explicit():
+        prop = model.JsonProperty(
+            name="json-val",
+            compressed=True,
+            json_type=tuple,
+            indexed=False,
+            repeated=False,
+            required=True,
+            default=(),
+            choices=((), ("b",), ("c", "d")),
+            validator=TestProperty._example_validator,
+            verbose_name="VALUE FOR READING",
+            write_empty_list=False,
+        )
+        assert prop._name == b"json-val" and prop._name != "json-val"
+        assert prop._compressed
+        assert prop._json_type is tuple
+        assert not prop._indexed
+        assert not prop._repeated
+        assert prop._required
+        assert prop._default == ()
+        assert prop._choices == frozenset([(), ("b",), ("c", "d")])
+        assert prop._validator is TestProperty._example_validator
+        assert prop._verbose_name == "VALUE FOR READING"
+        assert not prop._write_empty_list
+
+    @staticmethod
+    def test__validate_no_type():
+        prop = model.JsonProperty(name="json-val")
+        assert prop._validate(b"any") is None
+
+    @staticmethod
+    def test__validate_correct_type():
+        prop = model.JsonProperty(name="json-val", json_type=list)
+        assert prop._validate([b"any", b"mini"]) is None
+
+    @staticmethod
+    def test__validate_incorrect_type():
+        prop = model.JsonProperty(name="json-val", json_type=dict)
+        with pytest.raises(TypeError):
+            prop._validate(14)
+
+    @staticmethod
+    def test__to_base_type():
+        prop = model.JsonProperty(name="json-val")
+        value = [14, [15, 16], {"seventeen": 18}]
```

Comment:
Consider throwing some extended characters in here to make sure the survive the roundtrip to ascii and back.

Reply:
:+1: